app/gui/src/controller/searcher.rs

```diff
@@ -374,13 +377,71 @@ impl ThisNode {
 #[derive(Copy, Clone, Debug)]
 #[allow(missing_docs)]
 pub enum Mode {
-    /// Searcher should add a new node at a given position. `source_node` is either a selected node
+    /// Searcher is working with a newly created node. `source_node` is either a selected node
     /// or a node from which the connection was dragged out before being dropped at the scene.
-    NewNode { position: Option<Position>, source_node: Option<ast::Id> },
+    NewNode { node_id: ast::Id, source_node: Option<ast::Id> },
     /// Searcher should edit existing node's expression.
     EditNode { node_id: ast::Id },
 }
 
+impl Mode {
+    /// Return the ID of the node used as target for the Searcher.
+    pub fn node_id(&self) -> ast::Id {
+        match self {
+            Mode::NewNode { node_id, .. } => *node_id,
+            Mode::EditNode { node_id } => *node_id,
+        }
+    }
+
+    fn with_new_node(
+        parameters: SearcherParams,
+        graph: &presenter::Graph,
+        graph_editor: &GraphEditor,
+        graph_controller: &controller::Graph,
+    ) -> FallibleResult<Self> {
+        /// The expression to be used for newly created nodes when initialising the searcher without
+        /// an existing node.
+        const DEFAULT_INPUT_EXPRESSION: &str = "Nothing";
+        let SearcherParams { input, source_node } = parameters;
+
+        let view_data = graph_editor.model.nodes.get_cloned_ref(&input);
+
+        let position = view_data.map(|node| node.position().xy());
+        let position = position.map(|vector| model::module::Position { vector });
+
+        let metadata = NodeMetadata { position, ..default() };
+        let mut new_node = NewNodeInfo::new_pushed_back(DEFAULT_INPUT_EXPRESSION);
+        new_node.metadata = Some(metadata);
+        new_node.introduce_pattern = false;
+        let created_node = graph_controller.add_node(new_node)?;
```

Comment:
I think creating new node should not be a part of "Mode" constructor, as "Mode" is rather a description according to docs. A data structure, not an object in terms of OOP.